Readme.txt
Portable Object Compiler (c) 1997-2019.  All Rights Reserved.


GETTING HELP
------------

For help and reporting bugs : send email to me (stes@telenet.be)

INSTALL
-------

Read the Install.txt file.

If this is a first-time installation, you have to start by installing a bootstrap package; do not install this package until you have installed a bootstrap package.  This can be either a binary package of the Portable Object Compiler, or in some cases when no binary "port" exists an objc-bootstrap.tar.gz source package.

LICENSE
-------

Read the License.txt file (the "GNU Library General Public License").

The license file explains the difference between,

 (a) "software using the Portable Object Compiler", and 
 (b) "software based upon the Portable Object Compiler".


LATEST VERSION
--------------

The latest bootstrap compilers and sources of the Portable Object Compiler
are available at:

	http://sunsite.unc.edu/pub/Linux/devel/lang/objc

Or from my home page at:

	http://users.pandora.be/stes/

WHAT'S INCLUDED
---------------

The "objc" distribution consists of the following parts :

	0. "VICI", interactive Objective-C debugger and interpreter

	1. "objc", the actual compiler

	Consists of driver script ("objc") and binary ("objc1").
	"objc1" is a precompiler for C.
	The binary "postlink" is used
	for preparing tables for runtime initialization.

	2. "objcrt", the Objective C Runtime Library for "objc"

	This is a modern, portable, "all C" Objective C runtime
	There's absolutely no assembly language needed, nor is there
	a dependency on stack layout, register allocation conventions etc.

	Most other runtimes, were derived from pre-"all C"-Stepstone 
	runtimes, so this is an important point of difference.

	3. "objpak", the "Object Pak" Objective C Collection Classes

	Simple, powerful set of classes.  Very portable.
	Can be used to develop with different compilers, then
	port with 'objc'.  Compatible with ICpak101, the collection class
	library described in Brad Cox book, and used by existing
	Objective-C packages that were developed with Stepstone objc.

	4. "cakit", the "Computer Algebra Kit" Objective C classes

	Small, concise, interface to a large set of classes for
	polynomial computing and arbitrary precision integer arithmetic.

	5. Source of compiler and "oclib", a set of Objective C classes for
        parsing C (and the Objective C extensions to C)
  
"objc" and "vici" are written in Objective C itself.  It's all
written in portable Objective C : any Objective C compiler
should be able to compile the "objc" compiler.  From time to time,
we try a configure with other Objective-C compilers, to ensure that our
sources are still compatible with e.g. Stepstone Objective C compiler.

NOT included is a C compiler : in order to be able to use "objc", you will
need to elsewhere find a compatible C compiler.  This should be no problem
since one of our goals - see below - is to make "objc" work with as many C
compilers as possible.  On UNIX systems, the (free) "gcc" compiler is a
possibility (not needed, but possible).  Another possibility, is to use
the (free) "lcc" compiler or the "tcc" TDF compiler system.


NICE FEATURES
-------------

	1.Easy to install or to modify.

	2.Works on many systems with the native cc, debugger, profiler etc.
	(Unix, Windows, Macintosh, Beos, OpenVMS etc. see Platforms.txt file)

	2a. Option for reference counted memory management (-refcnt).
	This uses the native malloc(), free() etc. but the compiler generates
	statements for keeping track of references.
	(Tested on a few platforms, such as IRIX 5.2 with the SGI malloc).

	3.Built-in possibility of tracing Objective C messages. (OBJCRTMSG)

	4.Straightforward "C" messenger; "inline cache" messenger.
	Forwarding C messenger (to support -doesNotUnderstand:).
	
	5.All classes get a +initialize message at start-up, rather than
	'each class receives a +initialize before it receives its first
	message'.

	6.Some support for translating Objective-C to Smalltalk (-st80 option)

	7.Automatic archiver.  Compiler generates code for classes to save
	and load objects to and from disk (for all instance variables of
	type "id").

	8.Option for Garbage Collection (using Boehm gc package).  Tested
	on some UNIXes and WIN32.   Option for reference counted memory
	management (doesn't require Boehm).

	9.Exception handling scheme (using Objective-C Blocks) that allows
	to specify a default handler to be executed.

	10.Supports dynamically loading Objective-C modules on Windows,
	FreeBSD, HP-UX, Linux, IRIX, Digital Unix etc.

	11.Has a switch for double indirection for Object identifiers (id as
	a handle instead of a pointer). (-become: method)

	12.Supports forwarding messages (-doesNotUnderstand: method)

        13.Support for Embedded SQL in Objective-C (Informix only for now)  

	14.Great system for experimentation with your own additions/extensions
	to Objective C !


WANT TO HELP?
-------------

If you like "objc", feel free to send e-mail to, "stes@pandora.be".
I'm especially interested in bug-fixes, of course.  And in ports
to machines that I don't have access to (e.g. CP/M, Atari).



Changes.txt

3.3.16  * Solaris change : objc uses /usr/lib/amd64 when compiling with -m64

3.3.15  * Solaris change install libraries in prefix/lib/amd64 for 64case

	  this is to avoid setting LD_LIBRARY_PATH in the shared library case
	  as /usr/lib/amd64/ or /lib/amd64 are in the default crle setup
	  (or the link /usr/lib/64 and /lib/64)

	* set the -m64 flag via EXTRA_MFLAGS

	* compile the shared libraries with -xcode=pic13 in 64bit case

3.3.14	* Fedora 29 __builtin_bswap16 fix (gcc 8.2.1)

	* move #define OBJCVERSION to config.h in src/objc (from objc1.m)

3.3.13	* FreeBSD 11.2 _Noreturn and _Alignof fix 

3.3.12  * FreeBSD 10.4 and 11.2 and DragonFly BSD sys/_types.h runetype.h
	  clang 3.4/clang 6.0/gcc 8 fixes

	* add __alignof as builtin (in addition to the 
	  already present __alignof__, __builtin_alignof and __ALIGNOF__)

	* add _Thread_local typequal for runetype.h
	  extern _Thread_local const _RuneLocale *_ThreadRuneLocale;

3.3.11	* for Solaris and RHEL, separate package "objclibs"
	  with the shared libraries (libobjcrt.so and libobjpak.so runtime)
	  those are not required currently by the compiler itself
	  applications that are built against those shared libs,
	  can have a dependency on the objclibs package (runtime) only

	* set EXTRA_MFLAGS=-fPIC in RHEL case for building shared libraries
	  (in the objclibs.spec file)

	* fix configure script to add --build-id to ldargs on RHEL
	  this is enforced by the RHEL rpmbuild for packaging the shared libs

	* add script to build IPS Solaris packages directly (without SVR4 pkg)

	* improved Solaris package that installs in /usr instead of /usr/local

	* rename OBJCDIRLIB=lib or lib64  to LIB64NAME=lib or lib64

	* fix configure to install manpages in /usr/share/man 
	  instead of /usr/man  (there is a soft 'link' on Solaris)
	  the configure script was using the old (BSD?) /usr/man path

3.3.10	* update manual.html for version 3.3.10

	* fix location objcdlso.ld dlfile case Linux 64bit (lib64 issue)

	* fix C preprocessor with -traditional-cpp broken on darwin10/gcc 4.2.1

/usr/include/secure/_stdio.h:46: error: syntax error in macro parameter list
/usr/include/secure/_stdio.h:53: error: syntax error in macro parameter list

	* add installshlib target to Makefile to install shared libraries
	  on Solaris (verified with SUN developerstudio 12.6)

	* add EXTRA_MFLAGS to Makefile's (on UNIX)
	  could be used as in: ./configure EXTRA_MFLAGS=<extra flag>

3.3.9	* fixes i686-apple-darwin17.7.0 LLVM version 9.1.0 (clang-902.0.39.1)

	* add -darwin option to objc1 to enable some keywords like macosx,ios
	  modify configure.in script to add -darwin switch
	  (reported by A.Lipinski frotz@sympatico.ca)

	* fix for secure/_stdio.h darwin17.7 has a:

	          int     (* _Nullable _close)(void *);
	  
	  define _Nullable as -gnu type qualifier
	  define _Nonnull type qualifier , in stdlib.h

                  int (* _Nonnull __compar)(const void *, const void *));

	* fix for __attribute__ __attribute__ tmpnam() in stdio.h darwin17.7

	* fix for timingsafe_bcmp in string.h
	  has __attribute__((availability(macosx,introduced=10.12.1))) 

	  posing problem of those constants 10.12.1 (release numbers)
	  current fix could be withdrawn ...
	
	* fix for system() in stdlib.h __WATCHOS_PROHIBITED __TVOS_PROHIBITED

	* use -DOBJC_BLOCKS -U__BLOCKS__ in objc driver script 

	* some documentation updates / fix DependencyGraph example
	  (example from the Brad Cox book had useless #undef modification)

3.3.8	* Solaris IPS (Image Packaging System) manifest support

	* add amd64-pc-solaris11.4 with GCC 7.3 to Platforms GCC / Solaris11.4

        * test / upgrade Solaris 11.3 to DeveloperStudio 12.6 (from 12.4)
          pkg install pkg:/developer/developerstudio-126/cc             

	* fix 'unreached statement' warning in Block.m for SUNPRO

	* add i686-apple-darwin10.8.0 to Platforms.txt

	* Apple <stdio.h> has a <secure/_stdio.h> with 
	  /usr/include/secure/_string.h:103: warning: definition of
  	  '__builtin___stpcpy_chk' 

	  add '__builtin___stpcpy_chk' support
	 -D_FORTIFY_SOURCE=0 can be used as alternative

3.3.7	* support Solaris 11.4 cc: Studio 12.6 
	  by default the compiler builds as i386-pc-solaris2.11 32 bit
	  configure for 64 bit with ./configure amd64-pc-solaris11.4 

	* add _Atomic typeword for Solaris 11.4 stddef.h
	  -D__STDC_NO_ATOMICS__ could be used as alternative

3.3.6   * some fixes for "./configure amd64-pc-solaris11.3"
	  (or ./configure x86_64-pc-solaris11.3)

	  this is the -m64 option to support building ELF64 on Solaris 11.3
	  with the Sun/SolarisStudio 12.4 compiler

          by default the compiler is built "generic" (ELF32) but can be
          used to compile code for -m64 (ELF64) or be configured as 64bit by
	  ./configure x86_64-pc-solaris11.3

3.3.5   * add objc.SlackBuild script for building a Slackware package

3.3.4   * compiler driver fixes for SUN studio C
	Motif compile (imake) on Solaris11 support -z, -x, -B, -M
	support options -xO4 -xbuiltin=%all -xlibmil -xstrconst -xpentium -xregs=no%frameptr -Xa  -z lazyload -z combreloc -xstrconst -xildoff -B direct -M/usr/lib/ld/map (pass through those options, so that imake compile options work)

	* gcc 5 fix use 'gcc -E -x c' as cpp driver instead of "gcc -E"
	  ie. use gcc in 'C' mode rather than their 'Objective-C' mode
	  alternative would be to rename .m to .c (as CPPWANTSDOTC=Y on IRIX)
	  both work, observed on gcc 5 issue with -O2 which caused
	  the -D__GNUC_STDC_INLINE__ issue for the bits/string.h header
	  (the attribute(gnu_inline) not being emitted in Objective-C
	  ie. difference in output when invoking gcc -O2 -E or gcc -O2 -E -x c
	  they generate different code

3.3.3	* compile on Slackware 14.2 with gcc 5.3 and clang 3.8
	  NOTE: set OBJCOPT to -D__GNUC_STDC_INLINE__ for bootstrap
          cfr. gcc 5.2 with dragonfly 4.4.1  -D__GNUC_STDC_INLINE__

	* fix builtin type __builtin_offsetof for Motif X compile of richtext


src/objc/objc1.m

/* 
 * Copyright (c) 1998-2019 David Stes
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published 
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <string.h>		/* strtok */
#include <stdlib.h>		/* getenv */
#include <assert.h>
#ifndef __OBJECT_INCLUDED__
#define __OBJECT_INCLUDED__
#include <stdio.h> /* FILE */
#include "Object.h" /* Stepstone Object.h assumes #import */
#endif
#include <ocstring.h>
#include <ordcltn.h>
#include <sequence.h>
#include "config.h"
#include "node.h"
#include "trlunit.h"
#include "options.h"
#include "util.h"
#include "symbol.h"

void printversion(void)
{
  printf("%s\n",OBJCVERSION);
  exit(0);
}

void printcopyright(void)
{
  printf("Portable Object Compiler %s (c) 1997-2019.\n",OBJCVERSION);
  printf("Distributed under the terms of the GNU LGPL.\n");
}

void unknownoption(char* arg)
{
  STR msg = "%s: unknown option %s\n";
  STR name = (o_cplus) ? "objcpls1" : "objc1";
  fprintf(stderr, msg, name, arg);
  exit(1);
}

void badarg(id option,id arg)
{
  STR msg = "%s: illegal argument %s for %s\n";
  STR name = (o_cplus) ? "objcpls1" : "objc1";
  fprintf(stderr, msg, name, [arg str], [option str]);
  exit(1);
}

BOOL isoption(char* s)
{
  return s != NULL && s[0] == '-';
}

id optionstok(id aCltn, char *s)
{
  char *p;
  char *delims = " \t\n\r";
  id buffer = [String str:s];

  p = strtok([buffer str], delims);

  while (p != NULL) {
    [aCltn add:[String str:p]];
    p = strtok(NULL, delims);
  }

  return aCltn;
}

id optsfromfile(id aCltn, FILE * f)
{
  char buf[BUFSIZ + 1];

  while (!feof(f)) {
    if (fgets(buf, BUFSIZ, f)) {
      /* this works 'cause strtok also deletes \n */
      optionstok(aCltn, buf);
    }
  }

  return aCltn;
}

/* this is mostly for MS-DOS cmd line length limitations */

id optsnamed(id aCltn, STR s)
{
  STR t;
  FILE *f;

  /* maybe there's such an environment variable */
  if ((t = getenv(s)))
    return optionstok(aCltn, t);

  /* maybe there's such a file */
  if ((f = fopen(s, "r"))) {
    aCltn = optsfromfile(aCltn, f);
    fclose(f);
    return aCltn;
  }
  /* nope */
  return nil;
}

id cmdlineopts(int argc, char *argv[])
{
  int i;
  id aCltn = [OrdCltn new];

  for (i = 1; i < argc; i++) {
    [aCltn add:[String str:argv[i]]];
  }

  return aCltn;
}

/*
 * Filename for first line tag output (source filename)
 * (which we want to be file.m instead of file.P)
 */

int setfilename(id option, id args)
{
  id arg = [args next];
  char *t = [arg str];

  if (isoption(t)) badarg(option, arg);
  o_srcfilename = t;

  return 0;
}

int setinitcall(id option, id args)
{
  id arg = [args next];
  char *t = [arg str];

  if (isoption(t)) badarg(option, arg);
  o_initcall = t;

  return 0;
}

int setbrowsedir(id option, id args)
{
  id arg = [args next];
  char *t = [arg str];

  if (isoption(t)) badarg(option, arg);
  o_browsedir = t;

  return 0;
}

int setmainfun(id option, id args)
{
  id arg = [args next];
  if (isoption([arg str])) badarg(option, arg);
  o_mainfun = [arg str];
  return 0;
}

int setbind(id option, id args)
{
  id arg = [args next];
  if (isoption([arg str])) badarg(option, arg);
  o_bind = [arg str];
  return 0;
}

/*
 * Make line max settable (works only for identifiers now).
 */

int setlinemax(id option, id args)
{
  id arg = [args next];
  if (isoption([arg str])) badarg(option, arg);
  o_linemax = [arg asInt];
  return 0;
}

/*
 * @ filename is understood to take options from "filename"
 * Maybe we should also allow @filename (without space)
 */

void setoptions(id aCltn);

int setoptsfromfile(id option, id args)
{
  id arg = [args next];

  if (isoption([arg str])) badarg(option, arg);
  setoptions(optsnamed([OrdCltn new], [arg str]));
  return 0;
}

/*
 * Builtin functions (a la __builtin_isfloat).
 */

int addbuiltintype(id option, id args)
{
  id arg = [args next];
  if (isoption([arg str])) badarg(option, arg);
  definebuiltintype([arg str]);
  return 0;
}

int addbuiltinfun(id option, id args)
{
  id arg = [args next];
  if (isoption([arg str])) badarg(option, arg);
  definebuiltinfun([arg str]);
  return 0;
}

void setoptions(id aCltn)
{
  id args, s;
  int filecount = 0;		/* optional inFile and outFile */
  BOOL checkoption = YES;	/* YES if filename can't begin with a dash */

  args = [aCltn eachElement];

  while ((s = [args next])) {
    char *t = [s str];
    if (!strcmp(t,"-quiet")) {
      ++o_quiet;
    } else if (!strcmp(t,"-z")) {
      ++o_outputcode; 
    } else if (!strcmp(t,"@")) {
      setoptsfromfile(s, args);
    } else if (!strcmp(t,"-u")) {
      o_buffered=0;
    } else if (!strcmp(t,"-version")) {
      o_version++;
    } else if (!strcmp(t,"-otb")) {
      o_otb++;
      o_shareddata = 0;
    } else if (!strcmp(t,"-gnu")) {
      o_gnu++;
    } else if (!strcmp(t,"-darwin")) {
      o_darwin++;
    } else if (!strcmp(t,"-vms")) {
      o_vms++;
    } else if (!strcmp(t,"-hpux")) {
      o_hpux++;
    } else if (!strcmp(t,"-sunm64")) {
      o_sunm64++;
    } else if (!strcmp(t,"-refcnt")) {
      o_refcnt++;
      o_selfassign = 0; /* because receiver of a message is not counted */
    } else if (!strcmp(t,"-noasm")) {
      o_enableasm = 0;
    } else if (!strcmp(t,"-nolonglong")) {
      o_llkeyw = 0;
    } else if (!strcmp(t,"-msdos")) {
#if 0
 /* there are still some problems with this code on LCC */
 /* temporarily disable it (with the goal to enable this later again) */
      o_duptypedefs++;
#endif
      o_msdos++;
    } else if (!strcmp(t,"-watcom")) {
      o_watcom++;
    } else if (!strcmp(t,"-comments")) {
      o_comments++;
    } else if (!strcmp(t,"-duptypedefs")) {
      o_duptypedefs++;
    } else if (!strcmp(t,"-ibmvac")) {
      o_ibmvac++;
    } else if (!strcmp(t,"-noSelTbl")) {
      o_seltranslation = 0;
    } else if (!strcmp(t,"-noCategories")) {
      o_categories = 0;
    } else if (!strcmp(t,"-noBlocks")) {
      o_blocks = 0;
    } else if (!strcmp(t,"-st80")) {
      o_st80++;
      o_nolinetags++;
      o_outputcode = 0; /* no C code */
    } else if (!strcmp(t,"-fwd")) {
      o_fwd++;
    } else if (!strcmp(t,"-noSelPtr")) {
      o_selptr = 0;
    } else if (!strcmp(t,"-noFwd")) {
      o_fwd = 0;
    } else if (!strcmp(t,"-noCache")) {
      o_cache = 0;
    } else if (!strcmp(t,"-noFiler")) {
      o_filer = 0;
    } else if (!strcmp(t,"-noSelfAssign")) {
      o_selfassign = 0;
    } else if (!strcmp(t,"-noNilRcvr")) {
      o_nilrcvr = 0;
    } else if (!strcmp(t,"-objc")) {
      o_gencode = 0;
    } else if (!strcmp(t,"-cplus")) {
      o_cplus++;
    } else if (!strcmp(t,"-inlinecache")) {
      o_inlinecache++;
    } else if (!strcmp(t,"-refBind")) {
      o_refbind++;
    } else if (!strcmp(t,"-export")) {
      setbind(s, args);
    } else if (!strcmp(t,"-dllexport")) {
      o_bind = "__declspec(dllexport)";
    } else if (!strcmp(t,"-traditional")) {		/* ignore */
    } else if (!strcmp(t,"-filename")) {
      setfilename(s, args);
    } else if (!strcmp(t,"-longlinetag")) {		/* compatib. */
    } else if (!strcmp(t,"-shortTags")) {
      o_tagformat = "# %d \"%s\"\n";
    } else if (!strcmp(t,"-init")) {
      setinitcall(s, args);
    } else if (!strcmp(t,"-browse")) {
      setbrowsedir(s, args);
      o_nolinetags++;
      o_outputcode = 0; /* no C code */
    } else if (!strcmp(t,"-noShared")) {
      o_shareddata = 0;
    } else if (!strcmp(t,"-main")) {
      setmainfun(s, args);
    } else if (!strcmp(t,"-linemax")) {
      setlinemax(s, args);
    } else if (!strcmp(t,"-builtinfunction")) {
      addbuiltinfun(s, args);
    } else if (!strcmp(t,"-builtintype")) {
      addbuiltintype(s, args);
    } else if (!strcmp(t,"-nolinetags")) {
      o_nolinetags++;
    } else if (!strcmp(t,"-oneperfile")) {
      o_oneperfile++;
    } else if (!strcmp(t,"-w")) {
      o_warnings = 0;
    } else if (!strcmp(t,"-wClassUsedAsType")) {
      o_warnclasstype = 0;
    } else if (!strcmp(t,"-wMissing")) {
      o_warnmissingmethods = 0;
    } else if (!strcmp(t,"-wTypeConflict")) {
      o_warntypeconflict = 0;
    } else if (!strcmp(t,"-wLocalInstance")) {
      o_warnlocalnst = 0;
    } else if (!strcmp(t,"-wUndefinedMethod")) {
      o_warnundefined = 0;
    } else if (!strcmp(t,"-wInterfaceNotFound")) {
      o_warnnotfound = 0;
    } else if (!strcmp(t,"-WLex")) {
      o_warnlex++;
    } else if (!strcmp(t,"-WFwd")) {
      o_warnfwd++;
    } else if (!strcmp(t,"-postlink")) {
      o_postlink++;
    } else if (!strcmp(t,"-checkbindfunction")) {
      o_checkbind++;
    } else if (!strcmp(t,"-debugInfo")) {
      o_debuginfo++;
    } else if (!strcmp(t,"-ppi")) {
      o_ppi++;
    } else if (!strcmp(t,"-")) {
      checkoption = 0;
    } else if (checkoption && isoption(t)) {
      unknownoption(t);
    } else if (filecount == 0) {
      o_infile = t;
      filecount++;
    } else if (filecount == 1) {
      o_outfile = t;
      filecount++;
    } else {
      unknownoption(t);
    }
  }
}

/*
 * Opening and closing input file.  We're not dealing with wrapping around
 * files (include)  (because we're processing cpp output)
 */

void openinfile(void)
{
  if (o_infile) {
    yyin = openfile(o_infile, "r");
  } else {
    /* stdin */
  }
}

void setfirstlinetag(void)
{
  /* initialize linetag mechanism */

  inlineno = 1;
  if (o_srcfilename) {
    infilename = [String str:o_srcfilename];
  } else {
    if (o_infile) {
      infilename = [String str:o_infile];
    } else {
      infilename = [String str:"-=stdin=-"];
    }
  }

  if (!o_nolinetags) {
    gl(inlineno,[infilename str]);
  }

  assert(trlunit != nil);
  [trlunit setmodversion:OBJC1VERSION];
  [trlunit setmodname:[infilename str]];

  if (o_vms) {
    definebuiltinfun("__VA_START_BUILTIN");
    definebuiltinfun("__IS_X_FLOAT_BUILTIN");
    definebuiltintype("__int64");
  }

  if (o_hpux) {
    definebuiltinfun("__va_start__");
    definebuiltinfun("__va_arg__");
    definebuiltintype("__va_list__");
    definebuiltintype("__fpreg");
    definebuiltintype("__float80");
  }

  if (o_darwin) {
  /* Apple darwin headers have stuff like __attribute__((availability(macosx,introduced=10.12.1)))__attribute__((availability(ios,introduced=10.1))) */
  /* this causes some warnings about undefined tokens like ios */
    definebuiltinvar("message"); 
    definebuiltinvar("ios"); 
    definebuiltinvar("tvos"); 
    definebuiltinvar("watchos"); 
    definebuiltinvar("swift"); 
    definebuiltinvar("macos"); 
    definebuiltinvar("macosx"); 
    definebuiltinvar("introduced"); 
    definebuiltinvar("unavailable"); 
  }

  if (o_gnu) {
    definebuiltinfun("__builtin_bswap16");
    definebuiltinfun("__builtin_saveregs");
    definebuiltinfun("__builtin_classify_type");
    definebuiltinfun("__builtin_isfloat");
    definebuiltinfun("__builtin_classof");
    definebuiltinfun("__alignof");
    definebuiltinfun("_Alignof");
    definebuiltinfun("__alignof__");
    definebuiltinfun("__ALIGNOF__"); /* Intel icc 6.0 header */
    definebuiltinfun("__builtin_alignof");
    definebuiltinfun("__builtin_next_arg");
/*  definebuiltinfun("__builtin_alloca"); Sun Solaris 11.1 regular function */
    definebuiltinfun("__builtin_memcpy");
    definebuiltinfun("__builtin_memset");
    definebuiltinfun("__builtin_fabs");
    definebuiltinfun("__builtin_fabsl");
    definebuiltinfun("__builtin_fabsf");
    definebuiltinfun("__builtin_constant_p");
    definebuiltinvar("__func__"); /* C99 standard */
    definebuiltinvar("__FUNCTION__"); /* old GCC */
    definebuiltinvar("__PRETTY_FUNCTION__");
    definebuiltinvar("_VA_FP_SAVE_AREA");
    definebuiltintype("__builtin_va_list"); /* IA-64 */
    definebuiltinfun("__builtin_stdarg_start");
    definebuiltinfun("__builtin_va_end");
    definebuiltinfun("__builtin_va_arg");
    /* Solaris 11.3 defines va_start as function when -m64 */
    if (!o_sunm64) definebuiltinfun("__builtin_va_start");
    definebuiltinfun("__builtin_expect");
    definebuiltinfun("__builtin_strchr");
    definebuiltinfun("__builtin_strcmp");
    definebuiltinfun("__builtin_strncmp");
    definebuiltinfun("__builtin_offsetof"); /* Slackware 14.2 gcc 5.3 */
    definebuiltinfun("__builtin_strncpy"); /* CentOS 6.5 gcc 4.4.7 */

    /* gcc 4.5 for ubuntu 11.04 port */
    definebuiltinfun("__builtin_strlen");
    definebuiltinfun("__builtin_object_size");
    definebuiltinfun("__builtin_va_arg_pack");
    definebuiltinfun("__builtin___sprintf_chk");
    definebuiltinfun("__builtin___vsprintf_chk");
    definebuiltinfun("__builtin___snprintf_chk");
    definebuiltinfun("__builtin___vsnprintf_chk");
    definebuiltinfun("__builtin___strcpy_chk");
    definebuiltinfun("__builtin___strncpy_chk");
    definebuiltinfun("__builtin___strcat_chk");
    definebuiltinfun("__builtin___strncat_chk");
    definebuiltinfun("__builtin___memmove_chk");
    definebuiltinfun("__builtin___memset_chk");
    definebuiltinfun("__builtin___memcpy_chk");

    /* Slackware 14.1 port gcc 4.8.2  */
    definebuiltinfun("__builtin_bswap32");
    definebuiltinfun("__builtin_bswap64");

    /* Apple i686-apple-darwin10.8.0 */
    /* /usr/include/secure/_string.h:103 */
    definebuiltinfun("__builtin___stpcpy_chk");

    /* the following are not really gnu, they're SUN headers */
    /* Solaris 11.3 defines va_arg_incr as function when -m64 */
    definebuiltinvar("__builtin_va_alist");
    if (!o_sunm64) definebuiltinfun("__builtin_va_arg_incr");
  }
}

void closeinfile(void)
{
  if (o_infile) {
    fclose(yyin);
  } else {
    /* stdin */
  }
}

void openoutfile(void)
{
  if (o_outfile) {
    gfile = openfile(o_outfile,"w");
  } else {
    gfile = stdout; /* direct output to this stream */
  }

  /* for debugging output (-u unbuf option) */
  if (!o_buffered) setbuf(gfile,NULL);
}

void closeoutfile(void)
{
  if (o_outfile) fclose(gfile);
}

int main(int argc, char *argv[])
{
  /* define early since setoptions() modifies this object */
  trlunit = [TranslationUnit new];

  defoptions();
  setoptions(cmdlineopts(argc, argv));

  if (o_version) printversion();
  if (!o_quiet) printcopyright();

  openinfile();
  openoutfile(); /* redirect stdout */

  setfirstlinetag();

  if (o_outputcode) [trlunit prologue];
  if (yyparse()) exit(1);
  if (o_outputcode) [trlunit epilogue];

  closeinfile();
  closeoutfile();

  if (o_browsedir) [trlunit browse];

  exit(exitstatus);
  return 0;
}

